### PR TITLE
CI on GKE: cmd/test-pr

### PR DESCRIPTION
**Pull request by spxtr**
`cmd/test-pr` does a git clone, merges the PR, tars it and uploads it to GCS, and then reads `.test.yml` in the repo to decide what tests to run. Currently it does not actually run the tests.

I have a test for `checkoutPR` since that one is a bit tricky. It's not a hermetic test, but it should be well-behaved.

### CODE REVIEW

**Comment by spxtr** [comment]
ping @rmmh @ixdy cc @krousey

File: ciongke/cmd/test-pr/main.go [hunk 1]
```diff
@@ -0,0 +1,207 @@
...
+		RepoURL:  *repoURL,
+		RepoName: *repoName,
+		PRNumber: *pr,
+		Branch:   *branch,
+		Head:     *head,
+
+		Workspace:    *workspace,
+		SourceBucket: *sourceBucket,
+
+		GitHubClient: githubClient,
```
Comment (krousey) [inline ciongke/cmd/test-pr/main.go]: Looking through this, I can see where `KubeClient` will eventually be used, but I don't see anything that needs or will need `GitHubClient`.

**Comment by krousey** [comment]
Since I missed most of the other review, I thought I would mention that on the long-lived server, you might want to write your own [`oauth2.TokenSource`](https://godoc.org/golang.org/x/oauth2#TokenSource) that does a file watch on the secrets. That way you can update them without having to restart anything.

File: ciongke/cmd/test-pr/main.go (hunk 1 above)
Comment (spxtr) [inline ciongke/cmd/test-pr/main.go]: Oh right. I used to use that but I cut it out and forgot to remove it. Fixed, thanks.

**Comment by spxtr** [comment]
Oh neat, I didn't know that was possible. I'll add a TODO.

File: ciongke/gcs/gcs.go [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+type Client interface {
+	Upload(r io.Reader, bucket, name string) error
+}
+
+type client struct {
+	Service *storage.Service
+}
+
+// NewClient creates a client from within the cluster.
+func NewClient() (Client, error) {
```
Comment (krousey) [inline ciongke/gcs/gcs.go]: Don't return an interface. Drop the interface completely and just return the struct. If consumers need an interface, let them define the narrowest possible interface they need.

**Comment by krousey** [comment]
One comment, but otherwise looks good.

File: ciongke/gcs/gcs.go (hunk 2 above)
Comment (spxtr) [inline ciongke/gcs/gcs.go]: Should I do the same for the kubernetes client from the previous PR?

Comment (krousey) [inline ciongke/gcs/gcs.go]: I probably would, but in a separate PR.

File: ciongke/cmd/hook/main.go [hunk 3]
```diff
@@ -82,6 +82,8 @@ func main() {
 	}
 	webhookSecret := bytes.TrimSpace(webhookSecretRaw)
 
+	// TODO: Watch this file so that we don't need to manually restart when
+	// we update the token.
```
Comment (rmmh) [inline ciongke/cmd/hook/main.go]: Github only expire when revoked, so this seems like an unlikely problem.

File: ciongke/gcs/gcs.go (hunk 2 above)
Comment (spxtr) [inline ciongke/gcs/gcs.go]: Done.

**Comment by spxtr** [comment]
Thanks for the review :)

**Pull request merged**